README.md
# Nexmo JWT Generator for Ruby

[![Gem Version](https://badge.fury.io/rb/nexmo-jwt.svg)](https://badge.fury.io/rb/nexmo-jwt)![Coverage Status](https://github.com/nexmo/nexmo-jwt-ruby/workflows/CI/badge.svg)[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-v2.0%20adopted-ff69b4.svg)](CODE_OF_CONDUCT.md)

<img src="https://developer.nexmo.com/assets/images/Vonage_Nexmo.svg" height="48px" alt="Nexmo is now known as Vonage" />

This is the Ruby library to generate Nexmo JSON Web Tokens (JWTs). To use it you'll
need a Nexmo account. Sign up [for free at nexmo.com][signup].

* [Requirements](#requirements)
* [Installation](#installation)
* [Usage](#usage)
* [Documentation](#documentation)
* [License](#license)

## Requirements

The JWT generator supports Ruby version 2.5 or newer.

## Installation

To install the Ruby client library using Rubygems:

    gem install nexmo-jwt

Alternatively you can clone the repository:

    git clone git@github.com:Nexmo/nexmo-jwt-ruby.git

## Usage

By default the Nexmo JWT generator creates a short lived JWT (15 minutes) per request.
To generate a long lived JWT for multiple requests, specify a longer value in the `exp`
parameter during initialization. 

Example with no custom configuration:

```ruby
@builder = Nexmo::JWTBuilder.new(application_id: YOUR_APPLICATION_ID, private_key: YOUR_PRIVATE_KEY)
@token = @builder.jwt.generate
```

Example providing custom configuration options:

```ruby
@builder = Nexmo::JWTBuilder.new(
  application_id: YOUR_APPLICATION_ID,
  private_key: YOUR_PRIVATE_KEY,
  ttl: 500,
  paths: {
    "acl": {
      "paths": {
        "/messages": {
          "methods": ["POST", "GET"],
          "filters": {
            "from": "447977271009"  
          }     
        }  
      }   
    }
  },
  subject: 'My_Custom_Subject'
)
@token = @builder.jwt.generate
```

## Documentation

Nexmo Ruby JWT documentation: https://www.rubydoc.info/github/nexmo/nexmo-jwt

Nexmo Ruby code examples: https://github.com/Nexmo/nexmo-ruby-code-snippets

Nexmo API reference: https://developer.nexmo.com/api

## License

This library is released under the [MIT License][license]

[signup]: https://dashboard.nexmo.com/sign-up?utm_source=DEV_REL&utm_medium=github&utm_campaign=ruby-client-library
[license]: LICENSE.txt


CHANGES.md
# 0.1.0

First version released!

# 0.1.1

Fixed bug in setting of `exp` parameter and preserved ability to set it directly for backward
compability with the Nexmo Ruby SDK.

spec/nexmo-jwt/test.rb
# typed: ignore
require 'simplecov'

SimpleCov.start do
  add_filter 'spec/nexmo-jwt'
end

require 'minitest/autorun'
require_relative '../../lib/nexmo-jwt'

module NexmoJwt
  class Test < Minitest::Test
    def setup
      super

      @request_stubs = []
    end

    def stub_request(*args)
      super.tap do |stub|
        @request_stubs << stub
      end
    end

    def teardown
      @request_stubs.each do |stub|
        assert_requested(stub, at_least_times: 1)
      end

      super
    end

    def application_id
      'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
    end

    def private_key
      File.read(File.expand_path(File.join(File.dirname(__FILE__), '..', 'private_key.txt')))
    end
  end
end


spec/nexmo-jwt/spec_helper.rb
require 'bundler/setup'
Bundler.setup

require 'nexmo-jwt'

lib/nexmo-jwt/generator.rb
# typed: strict
require 'securerandom'
require 'openssl'
require 'jwt'

module NexmoJwt
  class Generator
    extend T::Sig

    # Generate an encoded JSON Web Token.
    #
    # By default the Nexmo JWT generator creates a short lived (15 minutes) JWT per request.
    #
    # To generate a long lived JWT for multiple requests, specify a longer value in the `exp`
    # parameter in the `claims` object. 
    #
    # @example
    #   claims = {
    #     application_id: application_id,
    #     nbf: 1483315200,
    #     exp: 1514764800,
    #     iat: 1483228800
    #   }
    #
    #   private_key = File.read('path/to/private.key')
    #
    #   token = NexmoJWT::Generator.generate(claims, private_key)
    #
    # @param [Hash] payload
    # @param [String, OpenSSL::PKey::RSA] private_key
    #
    # @return [String]
    #
    sig {params(
      payload: {
        "application_id" => T.nilable(String),
        "iat" => T.nilable(Integer),
        "exp" => T.nilable(Integer),
        "jti" => T.nilable(String)
      },
      private_key: OpenSSL::PKey::RSA
    ).returns(String)}
    def self.generate(payload, private_key)
      payload[:iat] = iat = Time.now.to_i unless payload.key?(:iat) || payload.key?('iat')
      payload[:exp] = T.must(iat) + 60 unless payload.key?(:exp) || payload.key?('exp')
      payload[:jti] = SecureRandom.uuid unless payload.key?(:jti) || payload.key?('jti')

      private_key = OpenSSL::PKey::RSA.new(private_key) unless private_key.respond_to?(:sign)

      ::JWT.encode(payload, private_key, 'RS256')
    end
  end
end

lib/nexmo-jwt/jwt_builder.rb
require 'openssl'
require 'securerandom'
require_relative 'jwt'

module Nexmo
  class JWTBuilder
    # Generate an encoded JSON Web Token.
    #
    # By default the Nexmo JWT generator creates a short lived (15 minutes) JWT per request.
    #
    # To generate a long lived JWT for multiple requests, specify a longer value in the `exp`
    # parameter during initialization. 
    #
    # Example with no custom configuration:
    #
    # @example
    #   @builder = Nexmo::JWTBuilder.new(application_id: YOUR_APPLICATION_ID, private_key: YOUR_PRIVATE_KEY)
    #   @token = @builder.jwt.generate
    #
    # Example providing custom configuration options:
    #
    # @example
    #   @generator = Nexmo::JWTBuilder.new(
    #     application_id: YOUR_APPLICATION_ID,
    #     private_key: YOUR_PRIVATE_KEY,
    #     ttl: 500,
    #     subject: 'My_Custom_Subject'
    #   )
    #   @token = @builder.jwt.generate
    #
    # @param [String] application_id
    # @param [Integer] iat
    # @param [SecureRandom::UUIDv4] jti
    # @param [Integer] nbf
    # @param [Integer] exp
    # @param [Hash] paths
    # @param [String] sub
    # @param [String, OpenSSL::PKey::RSA] private_key
    #
    # @return [String]
    #
    attr_accessor :application_id, :private_key, :jti, :nbf, :ttl, :exp, :alg, :paths, :subject, :jwt

    def initialize(params = {})
      Nexmo::JWTBuilder.validate_parameters_not_conflicting(params)

      @application_id = set_application_id(params.fetch(:application_id))
      @private_key = set_private_key(params.fetch(:private_key))
      @jti = params.fetch(:jti, SecureRandom.uuid)
      @nbf = params.fetch(:nbf, nil)
      @ttl = params.fetch(:ttl, 900)
      @exp = params.fetch(:exp, nil)
      @alg = params.fetch(:alg, 'RS256')
      @paths = params.fetch(:paths, nil)
      @subject = params.fetch(:subject, 'Subject')
      @jwt = Nexmo::JWT.new(:generator => self)

      after_initialize!(self)
    end

    def self.validate_parameters_not_conflicting(params)
      if params[:ttl] && params[:exp]
        raise ArgumentError, "Expected either 'ttl' or 'exp' parameter, preference is to set 'ttl' parameter"
      end
    end

    def after_initialize!(builder)
      validate_not_before(builder.nbf) if builder.nbf
      validate_time_to_live(builder.ttl)
      validate_paths(builder.paths) if builder.paths
      validate_subject(builder.subject) if builder.subject
    end

    def set_application_id(application_id)
      validate_application_id(application_id)

      application_id
    end

    def set_private_key(private_key)
      validate_private_key(private_key)

      if File.exist?(private_key)
        key = OpenSSL::PKey::RSA.new(File.read(private_key))
      else
        key = OpenSSL::PKey::RSA.new(private_key)
      end

      key
    end

    def set_exp
      Time.now.to_i
    end

    def validate_application_id(application_id)
      raise ArgumentError, "Missing required 'application_id' parameter" if application_id.nil?
    end

    def validate_private_key(private_key)
      raise ArgumentError, "Missing required 'private_key' parameter" if private_key.nil?
    end

    def validate_not_before(nbf)
      raise ArgumentError, "Expected Integer parameter type for NotBefore 'nbf' parameter" unless nbf.is_a?(Integer)
    end

    def validate_time_to_live(ttl)
      raise ArgumentError, "Expected Integer parameter type for TimeToLive 'ttl' parameter" unless ttl.is_a?(Integer)
    end

    def validate_paths(acl_paths)
      raise ArgumentError, "Expected Hash parameter type for Paths 'paths' parameter" unless acl_paths.is_a?(Hash)
    end

    def validate_subject(subject)
      raise ArgumentError, "Expected String parameter type for Subject 'subject' parameter" unless subject.is_a?(String)
    end
  end
end